language/translation/translation_training_data_tsv_generator.ipynb

```diff
@@ -0,0 +1,786 @@
+   "id": "de7fc4fe-50dd-439a-931d-aed77ded7053",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Generated TSV stored at /home/jupyter/src/your_tsv_file_name.tsv\n"
+     ]
+    }
+   ],
+   "source": [
+    "mismatched_block, more_than_200_words = make_tsv(source_ref_dictionary, tsv_file_name)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 73,
+   "id": "040d325e-9ef8-47fd-9bf7-b11c68f208d5",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{}"
+      ]
+     },
+     "execution_count": 73,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "mismatched_block"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 74,
+   "id": "51ceb169-ecb0-401d-ab6d-07aa1ebd3b77",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "{}"
+      ]
+     },
+     "execution_count": 74,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "more_than_200_words"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8c4bb7f5-4df1-445d-b97f-b026ed98ad92",
+   "metadata": {},
+   "source": [
+    "## Custom model training"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 85,
+   "id": "a89cbbda-8996-447e-abec-96de3c8bc71c",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "def generate_access_token() -> Credentials:\n",
+    "    \"\"\"Generates access token to call translate APIs.\"\"\"\n",
+    "    creds, project = google.auth.default()\n",
+    "\n",
+    "    auth_req = google.auth.transport.requests.Request()\n",
+    "    creds.refresh(auth_req)\n",
+    "    return creds.token\n",
+    "\n",
+    "\n",
+    "def create_dataset(\n",
+    "    target_lang_code: str,\n",
+    "    url: str,\n",
+    "    source_lang_code: Optional[str] = DEFAULT_SOURCE_LANG_CODE,\n",
+    ") -> dict or None:\n",
+    "    \"\"\"Creates a dataset.\"\"\"\n",
+    "    ACCESS_TOKEN = generate_access_token()\n",
+    "    headers = {\n",
+    "        \"Authorization\": f\"Bearer {ACCESS_TOKEN}\",\n",
+    "        \"Content-Type\": \"application/json; charset=UTF-8\",\n",
+    "    }\n",
+    "\n",
+    "    if DEFAULT_DATASET_SUFFIX != \"\" and DEFAULT_DATASET_SUFFIX is not None:\n",
+    "        dataset_display_name = f\"{DEFAULT_DATASET_PREFIX}_{source_lang_code}_to_{target_lang_code}_{DEFAULT_DATASET_SUFFIX}\"\n",
+    "    else:\n",
+    "        dataset_display_name = (\n",
+    "            f\"{DEFAULT_DATASET_PREFIX}_{source_lang_code}_to_{target_lang_code}\"\n",
+    "        )\n",
+    "\n",
+    "    data = {\n",
+    "        \"display_name\": dataset_display_name,\n",
+    "        \"source_language_code\": source_lang_code,\n",
+    "        \"target_language_code\": target_lang_code,\n",
```

Comment:
Thank you for adding this code to create the Dataset/Train the model, but please use the Translation API Client Library where possible.

https://cloud.google.com/translate/docs/reference/libraries/v3/python

Reply:
That is True. But  they don't have support for all actions via client Library API. I could only find support to List Glossaries, detect languages etc.

Reply:
I am unable to locate the API libraries for Creating dataset or importing data or training a model. Only Create Glossary, and translating text or document is available as API reference. Rest all we need to via REST APIs. And I think we need to generate access token for that. Let me know if this is not correct understanding. Thank you Holt !

Reply:
I think these are the methods you need to use:

- https://cloud.google.com/python/docs/reference/translate/latest/google.cloud.translate_v3.services.translation_service.TranslationServiceClient#google_cloud_translate_v3_services_translation_service_TranslationServiceClient_create_adaptive_mt_dataset
- https://cloud.google.com/python/docs/reference/translate/latest/google.cloud.translate_v3.services.translation_service.TranslationServiceClient#google_cloud_translate_v3_services_translation_service_TranslationServiceClient_import_adaptive_mt_file
- https://cloud.google.com/python/docs/reference/translate/latest/google.cloud.translate_v3.services.translation_service.TranslationServiceClient#google_cloud_translate_v3_services_translation_service_TranslationServiceClient_adaptive_mt_translate
- https://cloud.google.com/python/docs/reference/translate/latest/google.cloud.translate_v3.services.translation_service.TranslationServiceClient#google_cloud_translate_v3_services_translation_service_TranslationServiceClient_adaptive_mt_dataset_path

Reply:
Wait, I just realized that you're using these methods which are different than Adaptive MT.

https://cloud.google.com/translate/docs/reference/rest/v3/projects.locations.datasets

It does seem that these methods are missing from the Client Library, that shouldn't be the case. I'm going to file a bug to see about fixing that.

Reply:
I'm resolving this for now. @abhijat-gupta I ccd you on the bug to the product dev team to see about the methods being added.